File: app/javascript/channels/chat_room_channel.js
import consumer from "./consumer"

const appChatRoom = consumer.subscriptions.create("ChatRoomChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    const chatMessages = document.getElementById('chat-messages');
    chatMessages.insertAdjacentHTML('beforeend', data['chat_message']);
  },

  speak: function(chat_message, chat_room_id) {
    return this.perform('speak', { chat_message, chat_room_id });
  }
});

if(/chat_rooms/.test(location.pathname)) {
  $(document).on('keydown', '.chat-room__message-form_textarea', function(e) {
    if (e.key === 'Enter') {
      const chat_room_id = $('textarea').data('chat_room_id');
      appChatRoom.speak(e.target.value, chat_room_id);
      e.target.value = '';
      e.preventDefault();
    }
  });
}

File: app/channels/chat_room_channel.rb
class ChatRoomChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_room_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    ChatMessage.create!(
      content: data['chat_message'],
      user_id: current_user.id,
      chat_room_id: data['chat_room_id']
    )

    ActionCable.server.broadcast 'chat_room_channel', { chat_message: data['chat_message'] }
  end
end
